File: App/Resources/AlgorithmDetails/deterministicbogosort/deterministicbogosort.java
import java.util.Arrays;

public class deterministicbogosort {
  public static boolean isSorted(int[] arr) {
    for (int i = 0; i < arr.length - 1; i++) {
      if (arr[i] > arr[i + 1]) {
        return false;
      }
    }
    return true;
  }

  public static boolean permutationSort(int[] arr, int depth, int n) {
    if (depth >= n - 1) {
      return isSorted(arr);
    }
    for (int i = n - 1; i > depth; i--) {
      if (permutationSort(arr, depth + 1, n)) {
        return true;
      }
      if ((n - depth) % 2 == 0) {
        int temp = arr[depth];
        arr[depth] = arr[i];
        arr[i] = temp;
      } else {
        int temp = arr[depth];
        arr[depth] = arr[n - 1];
        arr[n - 1] = temp;
      }
    }
    return permutationSort(arr, depth + 1, n);
  }

  public static void sort(int arr[]) {
    int n = arr.length;
    permutationSort(arr, 0, n);
  }

  public static void main(String[] args) {
    int[] array = new int[]{0, 39, 21, 62, 91, 14, 23};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}

File: Modules/BuiltInAlgorithms/Sources/DeterministicBogoSort.swift
import AlgorithmKit
import SortEngineKit

public struct DeterministicBogoSort: SortAlgorithm {
    public let id = AlgorithmID(rawValue: "deterministicbogosort")
    public let metadata = AlgorithmMetadata(
        displayName: "Deterministic Bogo Sort",
        category: .impractical,
        sizeRange: 4...8,
        stable: false,
        timeComplexity: ComplexityBounds(best: "O(n)", average: "O(n \\times n!)", worst: "O(n \\times n!)"),
        spaceComplexity: "O(1)",
        iconName: "die.face.4.fill"
    )
    public init() {}

    /// Already fully deterministic in ArrayV — no `randInt` in the Java source, true to its name.
    /// Heap's algorithm walking every permutation, same technique `BozoSort` already ported, just
    /// structured as forward recursion counting a `depth` up from 0 (checking sortedness once
    /// `depth` reaches `n - 1`) rather than `BozoSort`'s own `k` counting down from `n`.
    public func record(into engine: inout RecordingEngine) {
        let n = engine.count
        guard n > 1 else { return }

        func isSorted() -> Bool {
            for i in 1..<n where !engine.compare(i, i - 1) { return false }
            return true
        }

        func permutationSort(_ depth: Int) -> Bool {
            if depth >= n - 1 {
                return isSorted()
            }
            for i in stride(from: n - 1, through: depth + 1, by: -1) {
                if permutationSort(depth + 1) { return true }
                if (n - depth).isMultiple(of: 2) {
                    engine.swap(depth, i)
                } else {
                    engine.swap(depth, n - 1)
                }
            }
            return permutationSort(depth + 1)
        }

        _ = permutationSort(0)
    }
}
